src/Features/LanguageServer/Protocol/LspServices/LspServices.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition.Hosting;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer;
+
+internal class LspServices : IDisposable
+{
+    private readonly ImmutableDictionary<Type, Lazy<ILspService, LspServiceMetadataView>> _lazyLspServices;
+
+    /// <summary>
+    /// Gates access to <see cref="_servicesToDispose"/>.
+    /// </summary>
+    private readonly object _gate = new();
+    private readonly HashSet<IDisposable> _servicesToDispose = new(ReferenceEqualityComparer.Instance);
+
+    public LspServices(
+        ImmutableArray<Lazy<ILspService, LspServiceMetadataView>> mefLspServices,
+        ImmutableArray<Lazy<ILspServiceFactory, LspServiceMetadataView>> mefLspServiceFactories,
+        WellKnownLspServerKinds serverKind,
+        ImmutableArray<Lazy<ILspService, LspServiceMetadataView>> baseServices)
+    {
+        // Convert MEF exported service factories to the lazy LSP services that they create.
+        var servicesFromFactories = mefLspServiceFactories.Select(lz => new Lazy<ILspService, LspServiceMetadataView>(() => lz.Value.CreateILspService(this, serverKind), lz.Metadata));
+
+        var services = mefLspServices.Concat(servicesFromFactories);
+
+        // Make sure that we only include services exported for the specified server kind (or NotSpecified).
+        services = services.Where(lazyService => lazyService.Metadata.ServerKind == serverKind || lazyService.Metadata.ServerKind == WellKnownLspServerKinds.NotSpecified);
+
+        // Include the base level services that were passed in.
+        services = services.Concat(baseServices);
```

Comment:
Is there a reason not to just concat this together with `servicesFromFactories` above, and do one server kind check for the whole thing? Or at least perhaps a debug.assert that ensures the base services are all for the right server kinds? Seems like this could be a future bug farm.